ssl/s3_srvr.c

```diff
@@ -2761,8 +2761,24 @@ int ssl3_get_client_key_exchange(SSL *s)
 
         /* Get our certificate private key */
         alg_a = s->s3->tmp.new_cipher->algorithm_auth;
-        if (alg_a & SSL_aGOST01)
-            pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
+        if (alg_a & SSL_aGOST01) {
+            /*
+             * New GOST ciphersuites have SSL_aGOST01 bit too
+             */
+            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
+            if (pk == NULL) {
+                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
+            }
+            if (pk == NULL) {
+                pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
+            }
+        } else if (alg_a & SSL_aGOST12) {
+            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
+            if (pk == NULL) {
+                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
+            }
+        }
+ 
```

Comment:
Hello Matt!

On Mon, Sep 21, 2015 at 10:38 PM, mattcaswell <notifications@github.com>
wrote:

> In ssl/s3_srvr.c
> <https://github.com/openssl/openssl/pull/407#discussion_r40015358>:
>
> > +             * New GOST ciphersuites have SSL_aGOST01 bit too
> > +             */
> > +            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
> > +            if (pk == NULL) {
> > +                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
> > +            }
> > +            if (pk == NULL) {
> > +                pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
> > +            }
> > +        } else if (alg_a & SSL_aGOST12) {
> > +            pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
> > +            if (pk == NULL) {
> > +                pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
> > +            }
> > +        }
> > +
>
> As I understand it from what you sent me the spec says you can only use
> GOST01 keys for GOST01 ciphers. For GOST12 ciphers you can use either
> GOST01 or GOST12 keys - right? In addition you have set the GOST01 bit for
> all GOST12 ciphers. So this logic doesn't quite do that I think. Isn't it
> supposed to look more like this?
>
>         if (alg_a & SSL_aGOST12) {
>             pk = s->cert->pkeys[SSL_PKEY_GOST12_512].privatekey;
>             if (pk == NULL) {
>                 pk = s->cert->pkeys[SSL_PKEY_GOST12_256].privatekey;
>             }
>             if (pk == NULL) {
>                 pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
>             }
>         } else if (alg_a & SSL_aGOST01) {
>             pk = s->cert->pkeys[SSL_PKEY_GOST01].privatekey;
>         }
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/openssl/openssl/pull/407/files#r40015358>.
>

I asked my former colleagues to make extra checks.I will report as soon as
I get the result.

-- 
SY, Dmitry Belyavsky